### ISSUE TITLE
Provide option to reorder grid cells in fulldistributed::Triangulation?

### ISSUE THREAD

**Issue by kronbichler** [dealii/dealii#17749]
Together with @dominiktassilostill, we worked on a case of a mesh where grid cells are very poorly ordered, and got bad performance of matrix-free algorithms: We lost about a factor of 2 in operator evaluation, only caused because the DoFs have very poor locality from one element to the next.

In https://github.com/dealii/dealii/blob/88c382288c89e734b2f87361df4f53aaaf6b676f/source/distributed/tria.cc#L2001-L2012 we reorder the cells to get a good coarse cell locality, and I wonder if we should do something similar with other grids, especially the fullydistributed Triangulation and sequential triangulations. Of course, a user might have good reasons to use a particular ordering, so doing something heuristic like `SparsityTools::reorder_hierarchic` (which is good, but by no means optimal) should certainly not the only option. I would therefore suggest to have something available as option to `GridIn` or some related process early in the triangulation creation. For the fully distributed triangulation in particular, I'd like to run it on the triangulation description that I get after partitioning, but before filling the local version of the triangulation.

Any opinions?

**Comment by peterrum** [comment]
@kronbichler Could you remind me what `setup_coarse_cell_to_p4est_tree_permutation()` is exactly doing. It creates a list of new cell ids for the coarse cells (via `SparsityTools::reorder_hierarchic()`). These ids are used, within p4est, for creating a 1D space-filling curve, which is partitioned.

Is this all? I don't see that `coarse_cell_to_p4est_tree_permutation` somehow influences the way cells are traversed locally (let's assume that the mesh is not refined). Am I wrong? A simple test, indicates that the order is not influenced:

https://github.com/dealii/dealii/blob/30a1331ba437fcc5fd1e5581c2190a7989ef6c81/tests/fullydistributed_grids/copy_distributed_tria_04.cc#L56-L65

In https://github.com/dealii/dealii/pull/17752, I have introduced a simple function `TriangulationDescription::Description::reorder_coarse_grid()` that allows to locally renumber the coarse-grid cells. It only works `TriangulationDescription::Description` is used. However, it gives indications:
- one needs to update `coarse_cells` and `coarse_cell_index_to_coarse_cell_id` if `TriangulationDescription::Description` is used; this is the case if `TriangulationDescription::Utilities::create_description_from_triangulation()` is used;
- one can reorder cells manually if `Triangulation::create_triangulation(vertices, cells)` is used, however this influences the cells on whole coarse grid
- one could internally do some renumbering during `Triangulation::create_triangulation(vertices, cells)`
- `Triangulation::create_triangulation(vertices, cells)` is called by `Triangulation::create_triangulation(description)`; if `Triangulation::create_triangulation(vertices, cells)` does the renumbering internally, the information has to be handed back somehow to `Triangulation::create_triangulation(description)`

**Comment by kronbichler** [comment]
> Could you remind me what `setup_coarse_cell_to_p4est_tree_permutation()` is exactly doing. It creates a list of new cell ids for the coarse cells (via `SparsityTools::reorder_hierarchical()`). These ids are used, within p4est, for creating a 1D space-filling curve, which is partitioned. Is this all?

Yes, this should be all as far as I remember, it is used to prescribe the order of coarse cells that p4est gets to see, and thus influences the way p4est creates the partitions of a parallel triangulation. I was imprecise (and in fact did not check) whether or not it also influences the local order of cells we see on the deal.II side. I intuitively thought it would, but your code shows it does not.

I like the code in the PR. This looks like the direction I've been thinking along, and you were much quicker with the solution than I had come. :+1:  I agree that a good interface could benefit all triangulation interfaces. As I wrote, the initial intent is to have a good order in the cell traversal, because grids generated by grid generators, especially for large simplex meshes, can have really bad orders (as I've seen in at least three different contexts with very different code pipelines now).

In terms of interface, I think that we should not hardcode `reorder_hierarchical` because that function is good, but still a heuristic. It might make sense to let the user inject an ordering algorithm (or coarse cell ordering given as a vector of numbers), and use our hierarchical algorithm if nothing is given. For the work with the triangulation description, we might need a lambda function. But as I said, we can use the default numbering if nothing better is given.

When I ran your test case I got reminded that the order in `reorder_hierarchical` is very dependent on the initial order of the mesh (a fact that made me push a correction to #17747 to not break tests at that point), so we ought to improve that code. I wrote it as a quick fix in #760 to some problem I had with partitioning, but I think one can make it more robust by better tie breaking and possibly some weights.

**Comment by peterrum** [comment]
I guess the question is _who_ should to the renumbering. Should it be the user or the triangulation class?

From your comments, I read that users should have some flexibility regarding the type of reordering (e.g., via lambda function).

So option 1) would be that users manually (with the help of helper functions) renumber the cells and/or the cells in `TriangulationDescription::Description` before passing them to `Triangulation::create_triangulation()` and 2) would be that the `Triangulation` class does this depending on configuration/attached call-back functions.

If we think renumber is a good thing in all cases, I would suggest to do option 2).

Furthermore, I think we should remove the special treatment in `p::d::T`.

@kronbichler What do you think? What would be your preference?

**Comment by dominiktassilostill** [comment]
When doing the tests, I saw that the partitioning of the fullydistributed triangulation works fine with the original and the re-ordered grid.
I renumbered the grid before creating `TriangulationDescription::Description` which was quite slow (first read-in a serial grid, renumber the cells and create a new serial grid), so is it possible for the `parallel::fullydistributed::Triangulation::create_triangulation(description)` function to only reorder the local cells or would one then also need to update the description and communicate it?

**Comment by peterrum** [comment]
> so is it possible for the parallel::fullydistributed::Triangulation::create_triangulation(description) function to only reorder the local cells or would one then also need to update the description and communicate it?

https://github.com/dealii/dealii/pull/17752 is doing it locally.

Generally, you create a coarse grid by creating a tuple `(vertices, cells, subcells)`. `GridGenerator` and `GridIn` do this and pass the information to `Triangulation::create_triangulation(vertices, cells, subcells)`. The order in `cells` determines the globally unique index of each coarse cell. Since, the coarse grid is replicated on each process for `Triangulation`/`p:s:T`/`p:d:T` the indices are the same and local ones and global ones are identical.

In the case of `p:f:T`, processes only store locally-relevant coarse cells. Internally, there is a map between local and global cell indices. You can permute the local indices as you want (independently what other processes are doing), as long as the global indices stay the same.

**Comment by bangerth** [comment]
In the case of `p::d::T`, since p4est orders all cells in the parallel forest along a space-filling curve that traverses one tree after the other, we saw that when a coarse grid has a lot of cells, each partition may span the trees that come from several cells. If the order of cells is poorly chosen (as may come from a mesh generator), you end up with partitions that geographically consist of several islands within the domain, with large interfaces. As a consequence, we *always* reorder the incoming mesh with Cuthill-McKee to improve the locality is the partitioning.

If you have similarly good reasons to reorder cells, my suggestion would be to just always do it when creating the triangulation. This is the sort of thing that if you make it optional, 90% of people never find out and will just not do, resulting in unnecessarily slow codes.
